code/datums/components/uplink.dm

```diff
@@ -164,21 +164,29 @@
 			var/datum/uplink_item/I = uplink_items[category][item]
 			if(I.limited_stock == 0)
 				continue
-			if(I.restricted_roles.len && I.discounted == FALSE)
+			if(I.restricted_roles.len && !I.discounted)
 				var/is_inaccessible = TRUE
 				for(var/R in I.restricted_roles)
 					if(R == user.mind.assigned_role || debug)
 						is_inaccessible = FALSE
 				if(is_inaccessible)
 					continue
-			if(I.restricted_species && I.discounted == FALSE)
-				if(ishuman(user))
-					var/is_inaccessible = TRUE
-					var/mob/living/carbon/human/H = user
-					for(var/F in I.restricted_species)
-						if(F == H.dna.species.id || debug)
-							is_inaccessible = FALSE
-							break
+			if(I.restricted_species && !I.discounted && ishuman(user))
+				var/is_inaccessible = TRUE
+				var/mob/living/carbon/human/H = user
+				for(var/F in I.restricted_species)
+					if(F == H.dna.species.id || debug)
+						is_inaccessible = FALSE
+						break
+				if(is_inaccessible)
+					continue
+			if(I.restricted_roundstart_traits && !I.discounted && ishuman(user))
+				var/is_inaccessible = TRUE
+				var/mob/living/carbon/human/H = user
+				for(var/T in I.restricted_roundstart_traits)
+					if(HAS_TRAIT_FROM_ONLY(H, T, ROUNDSTART_TRAIT) || debug)
+						is_inaccessible = FALSE
+						break
```

Comment:
Check for human before both of them so you're not checking twice. 

Also, it'd be better to use something like `locate(H.dna.species.id) in I.restricted_species` instead of using for loop

Reply:
Hey I really like the idea of using locate() instead, but I have never seen someone give me a code example of it that actually works, I'll try refactoring it all with this later

Reply:
@TheFakeElon it worked for the exact example you showed, so I tried adding it for each check, but I accidentally broke it somehow :( and I gotta go on a business trip tomorrow morning and will be back wednesday night.